### PR TITLE
Add C interface tutorial

### PR DESCRIPTION
**Pull request by davidbeckingsale**
Should be viewable on ReadTheDocs: https://umpire.readthedocs.io/en/feature-um-257-c-tutorial/

### CODE REVIEW

**Comment by davidbeckingsale** [comment]
@mcfadden8 this should be ready for review. It includes tutorial examples for allocators, resources, and pools in C, as well as the recent updates and additions to the C/Fortran API.

I propose we then start a new branch to figure out the extra Fortran stuff on the roadmap.

File: examples/tutorial/c/tut_pool.c [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+int main(int argc, char** argv)
+{
+  umpire_resourcemanager rm;
+  umpire_resourcemanager_get_instance(&rm);
+
+  umpire_allocator allocator;
+  umpire_resourcemanager_get_allocator_by_name(&rm, "HOST", &allocator);
+
+  umpire_allocator pool;
+  umpire_resourcemanager_make_pool_allocator(&rm, "pool", allocator, 1024*512, 512, &pool);
```
Comment (mcfadden8) [inline examples/tutorial/c/tut_pool.c]: Do we need to add a comment describing the arguments?  Or is the definition for this function (and other C API functions documented elsewhere?

File: src/umpire/strategy/AllocationStrategy.cpp [hunk 2]
```diff
@@ -25,16 +25,16 @@ AllocationStrategy::AllocationStrategy(const std::string& name, int id) noexcept
...
+  return m_name;
 }
 
-std::string
-AllocationStrategy::getName() noexcept
+void
+AllocationStrategy::release()
 {
-  return m_name;
+  UMPIRE_LOG(Info, "AllocationStrategy::release in a no-op");
```
Comment (mcfadden8) [inline src/umpire/strategy/AllocationStrategy.cpp]:
```suggestion
  UMPIRE_LOG(Info, "AllocationStrategy::release is a no-op");
```

File: examples/tutorial/c/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,37 @@
+##############################################################################
+# Copyright (c) 2018, Lawrence Livermore National Security, LLC.
```
Comment (mcfadden8) [inline examples/tutorial/c/CMakeLists.txt]:
```suggestion
# Copyright (c) 2018-2019, Lawrence Livermore National Security, LLC.
```

File: examples/tutorial/c/tut_allocator.c [hunk 4]
```diff
@@ -0,0 +1,37 @@
+//////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2018, Lawrence Livermore National Security, LLC.
```
Comment (mcfadden8) [inline examples/tutorial/c/tut_allocator.c]:
```suggestion
// Copyright (c) 2018-2019, Lawrence Livermore National Security, LLC.
```

File: examples/tutorial/c/tut_resources.c [hunk 5]
```diff
@@ -0,0 +1,49 @@
+//////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2018, Lawrence Livermore National Security, LLC.
```
Comment (mcfadden8) [inline examples/tutorial/c/tut_resources.c]:
```suggestion
// Copyright (c) 2018-2019, Lawrence Livermore National Security, LLC.
```

File: docs/sphinx/tutorial.rst [hunk 6]
```diff
@@ -33,3 +34,19 @@ sections.
    tutorial/introspection.rst
    tutorial/typed_allocators.rst
    tutorial/replay.rst
+
+We also have a tutorial for the C interface to Umpire. Complete example
+listings are available, and will be compiled if you have configured Umpire with
+``-DENABLE_C_API=On``. 
+
```
[the remark this answers is not in the document]
Comment (davidbeckingsale) [reply docs/sphinx/tutorial.rst]: Good idea, I'll do that.

File: examples/tutorial/c/tut_allocator.c [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+#include "umpire/interface/umpire.h"
+
+#define SIZE 1024
+
+int main(int argc, char** argv) {
+  umpire_resourcemanager rm;
+  umpire_resourcemanager_get_instance(&rm);
+
+  umpire_allocator allocator;
+  umpire_resourcemanager_get_allocator_by_name(&rm, "HOST", &allocator);
```
[the remark this answers is not in the document]
Comment (davidbeckingsale) [reply examples/tutorial/c/tut_allocator.c]: It does return the `umpire_allocator`, but I think passing it is a ref is more C-like. I'm not sure how we can handle errors in the C code just yet.

File: examples/tutorial/c/tut_pool.c (hunk 1 above)
Comment (davidbeckingsale) [reply examples/tutorial/c/tut_pool.c]: I will do that.

File: examples/tutorial/c/tut_allocator.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (mcfadden8) [reply examples/tutorial/c/tut_allocator.c]: Okay thanks.  We can perhaps do that as a later project?

[the remark this answers is not in the document]
Comment (davidbeckingsale) [reply examples/tutorial/c/tut_allocator.c]: Agreed. I'll create a ticket.

File: docs/sphinx/tutorial/c/pools.rst [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+used to customize how data is obtained from the system.
+
+In this example, we will look at creating a pool that can fulfill requests for
+allocations of any size. To create a new ``umpire_allocator`` using the pooling
+algorithm:
+
+.. literalinclude:: ../../../examples/tutorial/tut_pool.c
+                    :lines: 24-28
+
+The two arguments are the size of teh initial block that is allocated, and the
```
Comment (mcfadden8) [inline docs/sphinx/tutorial/c/pools.rst]:
```suggestion
The two arguments are the size of the initial block that is allocated, and the
```

**Pull request merged**